examples/server/server.cpp

```diff
@@ -0,0 +1,794 @@
+        break;
+      }
+    }
+  }
+  if(tokens_completion[0] == assistant_tag_tokens[0]) {
+    bool execute_matching = true;
+    if(tokens_completion.size() > 1) { // if user tag had been tested
+      for(int i = 1;i < assistant_tag_tokens.size(); i++) {
+        if(i >= tokens_completion.size()) {
+          match_token = i;
+          break;
+        }
+        if(tokens_completion[i] == assistant_tag_tokens[i]) {
+          continue;
+        } else {
+          execute_matching = false;
+          break;
+        }
+      }
+    }
+    while(execute_matching) {
+      if(match_token == assistant_tag_tokens.size()) { // all assistant tag tokens matched, return empty inference
+        return "";
+      }
+      tkn = nextToken();
+      tokens_completion.push_back(tkn);
+      if(tkn == assistant_tag_tokens[match_token]) { // the token follow the sequence
+        match_token++;
+      } else if(match_token < assistant_tag_tokens.size()) { // no complete all user tag
+        break;
+      }
+    }
+  }
+  std::string result = "";
+  for(llama_token token : tokens_completion) {
+    result += llama_token_to_str(ctx, token);
+  }
+  return result;
+}
+
+void Llama::release() {
+  // TODO: Clean the context
+  // llama_free(ctx);
+}
+
+void server_print_usage(int /*argc*/, char ** argv, const gpt_params & params) {
+    fprintf(stderr, "usage: %s [options]\n", argv[0]);
+    fprintf(stderr, "\n");
+    fprintf(stderr, "options:\n");
+    fprintf(stderr, "  -h, --help            show this help message and exit\n");
+    fprintf(stderr, "  -s SEED, --seed SEED  RNG seed (default: -1, use random seed for < 0)\n");
+    fprintf(stderr, "  --memory_f32          use f32 instead of f16 for memory key+value\n");
+    fprintf(stderr, "  --keep                number of tokens to keep from the initial prompt (default: %d, -1 = all)\n", params.n_keep);
+    if (llama_mlock_supported()) {
+        fprintf(stderr, "  --mlock               force system to keep model in RAM rather than swapping or compressing\n");
+    }
+    if (llama_mmap_supported()) {
+        fprintf(stderr, "  --no-mmap             do not memory-map model (slower load but may reduce pageouts if not using mlock)\n");
+    }
+    fprintf(stderr, "  -ngl N, --n-gpu-layers N\n");
+    fprintf(stderr, "                        number of layers to store in VRAM\n");
+    fprintf(stderr, "  -m FNAME, --model FNAME\n");
+    fprintf(stderr, "                        model path (default: %s)\n", params.model.c_str());
+    fprintf(stderr, "  -host                 ip address to listen (default 0.0.0.0)\n");
+    fprintf(stderr, "  -port PORT            port to listen (default 8080)\n");
+    fprintf(stderr, "\n");
+}
+
+int main(int argc, char ** argv) {
+
+  // own arguments required by this example
+  gpt_params default_params;
+
+  gpt_params params;
+  params.model = "ggml-model.bin";
+
+  std::string hostname = "0.0.0.0";
```

Comment:
It may be better to only listen on 127.0.0.1 by default, exposing the server to the network can be dangerous.

Reply:
You're right. I will fix it.